File: k9mail-library/src/test/java/com/fsck/k9/mail/helpers/TestTrustedSocketFactory.java
package com.fsck.k9.mail.helpers;


import java.io.IOException;
import java.net.Socket;
import java.security.KeyManagementException;
import java.security.NoSuchAlgorithmException;
import java.security.cert.X509Certificate;

import com.fsck.k9.mail.MessagingException;
import com.fsck.k9.mail.ssl.TrustedSocketFactory;
import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLSocketFactory;
import javax.net.ssl.TrustManager;


public class TestTrustedSocketFactory implements TrustedSocketFactory {
    private final X509Certificate serverCertificate;


    public static TestTrustedSocketFactory newInstance() {
        X509Certificate serverCertificate = KeyStoreProvider.getInstance().getServerCertificate();
        return new TestTrustedSocketFactory(serverCertificate);
    }

    private TestTrustedSocketFactory(X509Certificate serverCertificate) {
        this.serverCertificate = serverCertificate;
    }

    @Override
    public Socket createSocket(Socket socket, String host, int port, String clientCertificateAlias)
            throws NoSuchAlgorithmException, KeyManagementException, MessagingException, IOException {

        TrustManager[] trustManagers = new TrustManager[] { new VeryTrustingTrustManager(serverCertificate) };

        SSLContext sslContext = SSLContext.getInstance("TLS");
        sslContext.init(null, trustManagers, null);

        SSLSocketFactory sslSocketFactory = sslContext.getSocketFactory();
        return sslSocketFactory.createSocket(
                socket,
                socket.getInetAddress().getHostAddress(),
                socket.getPort(),
                true);
    }
}

File: mail/testing/src/main/java/com/fsck/k9/mail/testing/security/TestTrustedSocketFactory.kt
package com.fsck.k9.mail.testing.security

import com.fsck.k9.mail.MessagingException
import com.fsck.k9.mail.ssl.TrustedSocketFactory
import java.io.IOException
import java.net.Socket
import java.security.KeyManagementException
import java.security.NoSuchAlgorithmException
import java.security.cert.X509Certificate
import javax.net.ssl.SSLContext
import javax.net.ssl.TrustManager

/**
 * A test trusted socket factory that creates sockets that trust only a predefined server certificate
 */
object TestTrustedSocketFactory : TrustedSocketFactory {

    private val serverCertificate: X509Certificate by lazy {
        KeyStoreProvider.instance.serverCertificate
    }

    @Throws(
        NoSuchAlgorithmException::class,
        KeyManagementException::class,
        MessagingException::class,
        IOException::class,
    )
    override fun createSocket(
        socket: Socket?,
        host: String,
        port: Int,
        clientCertificateAlias: String?,
    ): Socket {
        val trustManagers: Array<TrustManager> = arrayOf(VeryTrustingTrustManager(serverCertificate))

        val sslContext = SSLContext.getInstance("TLS").apply {
            init(null, trustManagers, null)
        }

        val sslSocketFactory = sslContext.socketFactory

        return if (socket == null) {
            sslSocketFactory.createSocket(host, port)
        } else {
            sslSocketFactory.createSocket(
                socket,
                socket.inetAddress.hostAddress,
                socket.port,
                true,
            )
        }
    }
}
